domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +208,340 @@ class TopicListController @Inject constructor(
-          val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
-          if (nextChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - lastCompletedChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
-              storyId,
-              topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              nextChapterSummary.name,
-              nextChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+          lastCompletedChapterProgress != null -> {
+            if (lastCompletedChapterProgress.explorationId
+              != story.chapterList.last().explorationId
+            ) {
+              val lastChapterSummary: ChapterSummary? = story.chapterList.find { chapterSummary ->
+                lastCompletedChapterProgress.explorationId == chapterSummary.explorationId
+              }
+              val nextChapterIndex = story.chapterList.indexOf(lastChapterSummary) + 1
+              val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
+              if (nextChapterSummary != null) {
+                val numberOfDaysPassed = (
+                  oppiaClock.getCurrentCalendar().timeInMillis -
+                    lastCompletedChapterProgress.lastPlayedTimestamp
+                  ) / ONE_DAY_IN_MS
+                val promotedStory = createPromotedStory(
+                  storyId,
+                  topic,
+                  completedChapterProgressList.size,
+                  story.chapterCount,
+                  nextChapterSummary.name,
+                  nextChapterSummary.explorationId
+                )
+                if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
+                  recommendedStoryBuilder.addRecentlyPlayedStory(promotedStory)
+                } else {
+                  recommendedStoryBuilder.addOlderPlayedStory(promotedStory)
+                }
+              }
             }
           }
         }
       }
+      recommendedActivityListBuilder.setRecommendedStoryList(recommendedStoryBuilder)
     }
-    if ((ongoingStoryListBuilder.olderStoryCount + ongoingStoryListBuilder.recentStoryCount) == 0) {
-      ongoingStoryListBuilder.addAllRecentStory(createRecommendedStoryList())
-    }
-    return ongoingStoryListBuilder.build()
   }
 
-  private fun createRecommendedStoryList(): List<PromotedStory> {
-    val recommendedStories = ArrayList<PromotedStory>()
+  private fun createRecommendedStoryList(
+    topicProgressList: List<TopicProgress>,
+    recommendedActivityListBuilder: RecommendedActivityList.Builder,
+    recommendedStoryBuilder: RecommendedStoryList.Builder
+  ): List<PromotedStory> {
```

Comment:
From the contract of this method, it seems like it should be returning a RecommendedStoryList. Can you update these methods to avoid passing in builders as much as possible? Similar to vars, they're making it quite difficult to follow the flow of state changing in these methods.